### PR TITLE
Configure Windows Service via Registry keys

### PR DESCRIPTION
**Pull request by alwaysbusy**
Currently the windows service has a hard-coded collection frequency and set of metrics which are reported. This allows both of those to be configured by Windows registry keys. In `HKEY_LOCAL_MACHINE\SOFTWARE\pcm\service` several options can be given. Registry was chosen as a good configuration method as it can be easily controlled via several administrative tools, does not require file location handling and can be easily shared/copied with `.reg` files.

`SampleRate` is a DWORD value which specifies the millisecond delay between data collections. This allows sampling to increase or decrease easily without recompilation.
`Collect{Core,Socket,Qpi}` are DWORD values which can be set to `0` to disable collection of a category of data, or `1` to enable. This allows different metrics to be collected without recompilation, as well as making it easier to add additional metrics to the service in the future without impacting on current collection behavior.

Both `SampleRate` and `Collect*` values are optional and default to their previous hard-coded value/behavior if not configured. Configuration changes are not picked up on the fly, but instead on startup only.

### CODE REVIEW

**Comment by opcm** [comment]
@ogbrugge please review

File: PCM-Service_Win/PCMService.h [hunk 1]
```diff
@@ -140,26 +171,74 @@ namespace PCMServiceNS {
...
-                baseTicksForIpcHash_.Add(s, gcnew PerformanceCounter(CountersCore, MetricCoreBaseIpc, s, false));
-                relFreqHash_.Add(s, gcnew PerformanceCounter(CountersCore, MetricCoreFreqRel, s, false));
-                baseTicksForRelFreqHash_.Add(s, gcnew PerformanceCounter(CountersCore, MetricCoreFreqNom, s, false));
-                thermalHeadroomHash_.Add(s, gcnew PerformanceCounter(CountersCore, MetricCoreHeadroom, s, false));
-                CoreC0StateResidencyHash_.Add(s, gcnew PerformanceCounter(CountersCore, MetricCoreResC0, s, false));
-                CoreC3StateResidencyHash_.Add(s, gcnew PerformanceCounter(CountersCore, MetricCoreResC3, s, false));
-                CoreC6StateResidencyHash_.Add(s, gcnew PerformanceCounter(CountersCore, MetricCoreResC6, s, false));
-                CoreC7StateResidencyHash_.Add(s, gcnew PerformanceCounter(CountersCore, MetricCoreResC7, s, false));
+				s = UInt32(i).ToString(); // For core counters we use just the number of the core
+				if (collectionInformation_->core)
```
Comment (ogbrugge) [inline PCM-Service_Win/PCMService.h]: Can you fix indentation here?

File: PCM-Service_Win/PCMService.h [hunk 2]
```diff
@@ -421,6 +493,10 @@ namespace PCMServiceNS {
         initonly String^ MetricSocketResC7 = gcnew String(L"package C7-state residency (%)");
 
         initonly String^ MetricQpiBand = gcnew String(L"QPI Link Bandwidth");
+
+        // Configuration values
+        const int sampleRate_;
+		const CollectionInformation^ collectionInformation_;
```
Comment (ogbrugge) [inline PCM-Service_Win/PCMService.h]: And here

File: PCM-Service_Win/PCMService.h [hunk 3]
```diff
@@ -282,65 +342,77 @@ namespace PCMServiceNS {
...
-                        ((PerformanceCounter^)baseTicksForIpcHash_[s])->RawValue = ticks >> 17;
-                        ((PerformanceCounter^)relFreqHash_[s])->RawValue = ticks >> 17;
-                        ((PerformanceCounter^)baseTicksForRelFreqHash_[s])->IncrementBy(refTicks >> 17);
-                        ((PerformanceCounter^)thermalHeadroomHash_[s])->RawValue = coreState.getThermalHeadroom();
-                        ((PerformanceCounter^)CoreC0StateResidencyHash_[s])->RawValue = __int64(100.*getCoreCStateResidency(0,oldCoreStates[i], coreState));
-                        ((PerformanceCounter^)CoreC3StateResidencyHash_[s])->RawValue = __int64(100.*getCoreCStateResidency(3,oldCoreStates[i], coreState));
-                        ((PerformanceCounter^)CoreC6StateResidencyHash_[s])->RawValue = __int64(100.*getCoreCStateResidency(6,oldCoreStates[i], coreState));
-                        ((PerformanceCounter^)CoreC7StateResidencyHash_[s])->RawValue = __int64(100.*getCoreCStateResidency(7,oldCoreStates[i], coreState));
-                        oldCoreStates[i] = coreState;
+						s = UInt32(i).ToString();
```
Comment (ogbrugge) [inline PCM-Service_Win/PCMService.h]: Indentation is broken here as well

File: PCM-Service_Win/PCMService.h [hunk 4]
```diff
@@ -482,6 +558,41 @@ namespace PCMServiceNS {
         /// </summary>
         virtual void OnStart(array<String^>^ args) override
         {
+            // Default values for configuration
+            int sampleRate = 1000;
+			CollectionInformation^ collectionInformation = gcnew CollectionInformation();
```
Comment (ogbrugge) [inline PCM-Service_Win/PCMService.h]: And here

File: PCM-Service_Win/PCMService.h [hunk 5]
```diff
@@ -482,6 +558,41 @@ namespace PCMServiceNS {
...
+
+                DWORD sampleRateRead(0);
+                if (ERROR_SUCCESS == RegQueryValueEx(hkey, L"sampleRate", NULL, NULL, reinterpret_cast<LPBYTE>(&sampleRateRead), &lenDWORD))
+                {
+                    sampleRate = (int)sampleRateRead;
+                }
+
+                DWORD collectCoreRead(0);
+                if (ERROR_SUCCESS == RegQueryValueEx(hkey, L"collectCore", NULL, NULL, reinterpret_cast<LPBYTE>(&collectCoreRead), &lenDWORD)) {
+					collectionInformation->core = (int)collectCoreRead > 0;
```
Comment (ogbrugge) [inline PCM-Service_Win/PCMService.h]: Here as well

File: PCM-Service_Win/PCMService.h [hunk 6]
```diff
@@ -482,6 +558,41 @@ namespace PCMServiceNS {
...
+                }
+
+				DWORD collectSocketRead(0);
```
Comment (ogbrugge) [inline PCM-Service_Win/PCMService.h]: Here as well

File: PCM-Service_Win/PCMService.h [hunk 7]
```diff
@@ -36,10 +36,30 @@ namespace PCMServiceNS {
         static initonly String^ ServiceName = gcnew String(L"PCMService");
     };
 
+	ref struct CollectionInformation {
```
Comment (ogbrugge) [inline PCM-Service_Win/PCMService.h]: Please indent at the same level as the class below

**Comment by alwaysbusy** [comment]
The latest commit should have fixed all indentation and casing issues. I upgraded my IDE just before working on this and only noticed after pushing that it hadn't kept my formatting settings.

**Comment by ogbrugge** [comment]
I unfortunately do not see those changes, not here and not in 145 or 146. Am I missing something?

**Comment by alwaysbusy** [comment]
Commit 755ac19 contains the white space changes, the spacing was only incorrect in the first commit to this branch (5242e57). On your comments above, GitHub is linking to the first commit only. A fresh, local checkout of the branch is showing everything correctly aligned for me.

**Comment by ogbrugge** [comment]
Indeed, I missed this change, approved now.

**Pull request merged**